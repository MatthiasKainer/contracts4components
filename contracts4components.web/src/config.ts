let __framework = "dom"

export const getFramework = () => __framework
export const setFramework = (framework: "dom" | "react") => __framework = framework