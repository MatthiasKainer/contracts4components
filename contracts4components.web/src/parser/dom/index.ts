import { parseGiven } from "./given";
import { parseThen } from "./then";
import { parseWhen } from "./when";

export {parseGiven, parseThen, parseWhen}