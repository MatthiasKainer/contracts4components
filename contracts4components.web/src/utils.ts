import RandExp from "randexp"

const capitalize = (s?: string) => `${s?.charAt(0).toUpperCase()}${s?.slice(1).toLowerCase()}`;

const isRequired = (name: string) => {
  throw new TypeError(`${name} parameter is required`);
};

const getValidText = (text: RegExp | undefined) =>
  text && text.test ? new RandExp(text).gen().toLocaleUpperCase() : (text?.toString() ?? "")

const parameterize = (arg: any) => {
  if (arg.test) {
    return arg
  }
  if (arg instanceof HTMLElement) {
    // naive assumption that every non-resolved instance is the container...
    return "container"
  }
  if ("string" === typeof arg) {
    return JSON.stringify(arg)
  }
  if ("object" === typeof arg) {
    return `{ ${Object.entries(arg).map(([key, value]) => `${key}: ${parameterize(value)}`).join(",")} }`
  }
  if ("function" === typeof arg) {
    return `(${arg.toString()})()`
  }

  return arg
}

const javaDateNow = () => {
  /*
  [
      2021,
      6,
      12,
      0,
      0
  ]
  */
  const date = new Date(Date.now())
  return [
    date.getUTCFullYear(),
    date.getUTCMonth() + 1,
    date.getUTCDay(),
    date.getUTCHours(),
    date.getUTCMinutes(),
    date.getUTCSeconds(),
    date.getUTCMilliseconds()
  ]
}

export { capitalize, isRequired, getValidText, parameterize, javaDateNow }