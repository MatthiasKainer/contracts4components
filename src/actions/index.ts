import {replaceActions} from "./replaceActions"
import userEvent from "@testing-library/user-event"

const {type, click} = userEvent

export const expectedAction = {
    ...replaceActions({
        type,
        click
    })
}