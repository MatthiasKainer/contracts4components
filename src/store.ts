import { writer } from "./writer"

const given: string[] = []
const when: string[] = []
const then: string[] = []

const normaliseLines = (line: string[]) => line.map(l => l.split("\n").map(l => l.trim())).reduce((prev, curr) => [...prev, ...curr], [])

export const cleanStore = () => {
    given.splice(0, given.length)
    when.splice(0, when.length)
    then.splice(0, then.length)
}

export const addGiven = (...line: string[]) => {
    given.push(...normaliseLines(line))
}
export const addWhen = (...line: string[]) => {
    when.push(...normaliseLines(line))
}
export const addThen = (...line: string[]) => {
    then.push(...normaliseLines(line))
}

export const toContract = () => {
    return writer().frame(...given,
        ...when,
        ...then)
}