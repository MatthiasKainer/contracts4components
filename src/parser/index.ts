import { Definition } from "../blocks/types";
import * as dom from "./dom"

export { GivenBlock, WhenStatement, UserActionStatement, QueryStatement, WhenBlock, ThenStatement, ThenEventListenerStatement, ThenBlock, Definition } from "../blocks/types";

export const parse = (container: HTMLElement, definition: Definition) => {
    dom.parseGiven(container, definition.given);
    dom.parseWhen(container, definition.when);
    dom.parseThen(container, definition.then);
}