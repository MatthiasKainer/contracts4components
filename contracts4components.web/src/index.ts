import { Definition, parse } from "./parser";
import { cleanStore, toContract } from "./store";

import * as fs from "fs"
import * as path from "path"
import * as vm from "vm"

import { setFramework } from "./config";
import { request } from "./net/request";
import { Contract } from "./domain/contract";
import { ContractIdentifier } from "./blocks/types";
import { javaDateNow } from "./utils";

export const setAsDOM = () => setFramework("dom")
export const setAsReact = () => setFramework("react")

type StoreOptions = {
    file?: string,
    url?: string
}

const test = (container: HTMLElement | any, definition: Definition, store: StoreOptions) => {
    return new Promise((resolve, reject) => {
        cleanStore()
        parse(container, definition)
        const name = `${definition.provider}-${definition.consumer}-${definition.element}`
        const promiseFile = new Promise((resolve, reject) => {
            if (store.file) {
                const location = path.join(store.file, `${name}.contract.js`)
                fs.writeFile(location, toContract(definition).join("\n"), (err) => {
                    if (err) return reject(err)
                    resolve(`${location} written`)
                })
            } else {
                resolve("no file provided")
            }
        })
        const promiseUrl = new Promise((resolve, reject) => {
            if (store.url) {
                const data = {
                    provider: definition.provider,
                    consumer: definition.consumer,
                    element: definition.element,
                    fileLines: toContract(definition)
                }
                request(store.url, { method: "POST", expectedStatus: 201 }, data)
                    .then(resolve)
                    .catch(reject)
            } else {
                resolve("no url provided")
            }
        })
        Promise.all([promiseFile, promiseUrl])
            .then(resolve)
            .catch(reject)
    })
}

export const testForDOM = (container: HTMLElement | any, definition: Definition, store: StoreOptions) => (setAsDOM(), test(container, definition, store))
export const testForReact = (container: HTMLElement | any, definition: Definition, store: StoreOptions) => (setAsReact(), test(container, definition, store))

export const verifyContract = async (container: HTMLElement | any, options: StoreOptions & ContractIdentifier & {version: string}) => {
    if (options.file) {
        await (require(options.file)).expectedContract(container)
    } else if (options.url) {
        const contracts = JSON.parse(await request(`${options.url}?provider=${options.provider}&consumer=${options.consumer}&element=${options.element}`, { method: "GET", expectedStatus: 200 })) as Contract[]
        if (contracts.length > 1) {
            throw new Error("More than one contract returned, combination of provider, consumer and element not unique.")
        }
        if (contracts.length < 1) {
            throw new Error("No contract found. Please provide an url to an existing contract.")
        }
        const [contract] = contracts
        const latestVersion = Object.values(contract.versions).pop()
        try {
            await eval(latestVersion.fileLines.join("\n")).expectedContract(container)

            await request(`${options.url}/${contract.id}/testResults`, { method: "PUT", expectedStatus: 202 }, {
                date: javaDateNow(),
                result: "Success",
                version: options.version
            })

        } catch (e) {
            await request(`${options.url}/${contract.id}/testResults`, { method: "PUT", expectedStatus: 202 }, {
                date: javaDateNow(),
                result: "Failure",
                version: options.version
            })
            throw e
        }
    }
}

export * from "./events"
export * from "./actions"
export * from "./queries"
export * from "./blocks/types"