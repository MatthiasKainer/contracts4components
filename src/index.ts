import { Definition, parse } from "./parser";
import { cleanStore, toContract } from "./store";

import * as fs from "fs"
import * as path from "path"
import { setFramework } from "./config";

export const setAsDOM = () => setFramework("dom")
export const setAsReact = () => setFramework("react")

const test = (container: HTMLElement | any, definition: Definition, store: string) => {
    return new Promise((resolve, reject) => {
        cleanStore()
        parse(container, definition)
        const location = path.join(store, `${definition.provider}-${definition.consumer}-${definition.element}.contract.js`)
        fs.writeFile(location, toContract().join("\n"), (err) => {
            if (err) return reject(err)
            resolve(`${location} written`)
        })
    })
}

export const testForDOM = (container: HTMLElement | any, definition: Definition, store: string) => (setAsDOM(), test(container, definition, store))
export const testForReact = (container: HTMLElement | any, definition: Definition, store: string) => (setAsReact(), test(container, definition, store))

export * from "./events"
export * from "./actions"
export * from "./queries"
export * from "./blocks/types"