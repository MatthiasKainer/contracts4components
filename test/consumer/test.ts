import { ThenBlock, ThenEventListenerStatement, UserActionStatement, WhenBlock } from "../../src/parser";
import {test} from "../../src"
import * as path from "path"
import { toContract } from "../../src/store";

describe("consumer <search molecule>", () => {

    it("provides a valid component", async () => {
        // given
        const searchQuery = { query: "some text" }
        const container = document.createElement("search-form")
        const definition = {
            provider: "search",
            consumer: "navigation",
            element: "search-form",
            given: {
                props: {
                    label: "my search"
                },
                events: {
                    "search": { detail: searchQuery }
                }
            },
            when: [
                {
                    type: "userAction",
                    action: "type",
                    args: [
                        {
                            type: "query",
                            action: "getByLabelText",
                            args: [
                                container,
                                /my search/gi
                            ]
                        },
                        searchQuery.query
                    ]
                },
                {
                    type: "userAction",
                    action: "click",
                    args: [
                        {
                            type: "query",
                            action: "getByRole",
                            args: [
                                container,
                                "button",
                                { name: /search/i }
                            ]
                        }
                    ],
                    triggers: "search"
                } as UserActionStatement
            ] as WhenBlock,
            then: [
                {
                    type: "eventListener",
                    name: "search",
                    assert: (search: jest.Mock) => {
                        expect(search).toBeCalledTimes(1);
                        expect(search.mock.calls[0][0].detail).toEqual(searchQuery);
                    }
                } as ThenEventListenerStatement
            ] as ThenBlock
        }
        
        expect(await test(container, definition, path.join(__dirname, ".."))).toMatchSnapshot()
        expect(toContract()).toMatchSnapshot()
        expect(container.outerHTML).toMatchSnapshot()
    })
})