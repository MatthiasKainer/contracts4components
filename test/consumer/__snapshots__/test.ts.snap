// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consumer <search molecule> provides a valid component 1`] = `"/Users/matthiaskainer/projects/private/contracts4components/test/search-navigation-search-form.contract.js written"`;

exports[`consumer <search molecule> provides a valid component 2`] = `
Array [
  "const {screen,getByRole,getByLabelText,getByText,} = require(\\"@testing-library/dom\\");",
  "const {default: userAction} = require(\\"@testing-library/user-event\\");",
  "module.exports = {",
  "expectedContract: async (container) => {",
  "container.setAttribute(\\"label\\", \\"my search\\");",
  "await new Promise((resolve) => setImmediate(resolve));",
  "const search = jest.fn();",
  "container.addEventListener(\\"search\\", search);",
  "const block_0_0 = screen.getByLabelText(/my search/gi);",
  "const block_0 = userAction.type(block_0_0,\\"some text\\");",
  "const block_1_0 = screen.getByRole(\\"button\\",{ name: /search/i });",
  "const block_1 = userAction.click(block_1_0);",
  "(function (search) {",
  "expect(search).toBeCalledTimes(1);",
  "expect(search.mock.calls[0][0].detail).toEqual({ query: \\"some text\\" });",
  "})(search)",
  "}",
  "}",
]
`;

exports[`consumer <search molecule> provides a valid component 3`] = `"<search-form label=\\"my search\\"><div><label for=\\"MY SEARCH\\">My search</label><input id=\\"MY SEARCH\\"></div><div role=\\"button\\">SEARCH</div></search-form>"`;
